### PR TITLE
Geod 503 null coordinates causes error

### PR DESCRIPTION
**Pull request by oehm-smith**
~~DON'T DO THIS UNTIL https://github.com/GeoscienceAustralia/GNSS-Site-Manager/pull/453/files is in~~

All ready to go

### CODE REVIEW

File: src/client/app/shared/json-data-view-model/data-view-translator.spec.ts [hunk 1]
```diff
@@ -4,10 +4,14 @@ import { JsonViewModelServiceSpecData } from './json-view-model.service.spec.dat
 import { HumiditySensorViewModel } from '../../humidity-sensor/humidity-sensor-view-model';
 import { GnssReceiverViewModel } from '../../gnss-receiver/gnss-receiver-view-model';
 import { MiscUtils } from '../global/misc-utils';
+import { DataViewTranslatorSpecData } from './data-view-translator.spec.data';
+import { SiteLocationMappings } from '../../site-log/site-location.mapping';
+import { JsonViewModelService } from './json-view-model.service';
```
Comment (lbodor) [inline src/client/app/shared/json-data-view-model/data-view-translator.spec.ts]: This is a circular dependency, which is a bad sign. Let's not fix it now, because both data-view-translator and json-view-model-service will be reworked.

File: src/client/app/shared/json-data-view-model/data-view-translator.ts [hunk 2]
```diff
@@ -71,6 +71,28 @@ export class DataViewTranslatorService {
...
+                    if (fieldMap.sourceField.pointer.match(/cartesianPosition/)) {
+                        return this.translateCartesianPosition(source, {viewToData: true});
+                    } else if (fieldMap.sourceField.pointer.match(/geodeticPosition/)) {
+                        return this.translateGeodeticPosition(source, {viewToData: true});
+                    } else {
+                        throw new Error(`DataViewTranslatorService - unknown sourceField.pointer: 
+                        ${fieldMap.sourceField.pointer} for sourceField.type: ${fieldMap.sourceField.type}`);
+                    }
+
+                }
```
Comment (lbodor) [inline src/client/app/shared/json-data-view-model/data-view-translator.ts]: If we can, let's do the point conversion after the check for `source === undefined` below.

File: src/client/app/shared/json-data-view-model/data-view-translator.ts [hunk 3]
```diff
@@ -92,10 +114,95 @@ export class DataViewTranslatorService {
...
+     * @param viewToDataTranslateOptions - {viewToData: boolean} - default is {viewToData: true}
+     * @return the translated value for the top-level CartesianPosition or GeodeticPosition
+     */
+    static translateCartesianPosition(source: any, viewToDataTranslateOptions?: { viewToData: boolean }): any {
+        let mapper = createMapper({alwaysTransform: true, alwaysSet: true});
+        if (viewToDataTranslateOptions
+            && viewToDataTranslateOptions.hasOwnProperty('viewToData')
+            && !viewToDataTranslateOptions['viewToData']) {
+            // data to view translate
+            if (! source || (typeof source === 'object' && !source.hasOwnProperty('point'))) {
```
Comment (lbodor) [inline src/client/app/shared/json-data-view-model/data-view-translator.ts]: No spae after !

File: src/client/app/shared/jsonix/jsonix.service.spec.ts [hunk 4]
```diff
@@ -287,5 +288,30 @@ export function main() {
...
+
+      });
   });
+
+    /**
+     * Rewrite the above tests using the one data file with modifications such as like that performed in this function.
+     * @param json test data
+     * @return json test data modified for the test
+     */
+    function modify_undefine_Location_geodeticPosition(json: any): any {
```
Comment (lbodor) [inline src/client/app/shared/jsonix/jsonix.service.spec.ts]: No underscores please. Verbs for functions are a good idea, but two is not better than one :)

File: src/client/app/shared/jsonix/jsonix.service.spec.ts [hunk 5]
```diff
@@ -287,5 +288,30 @@ export function main() {
...
+              expect(geodesyMl).toContain('<geo:approximatePositionITRF>');
+              expect(geodesyMl).toContain('<geo:cartesianPosition/>');
+              expect(geodesyMl).toContain('<geo:geodeticPosition><gml:Point><gml:pos>4 5 6</gml:pos></gml:Point></geo:geodeticPosition>');
+          });
+
+      });
   });
+
+    /**
+     * Rewrite the above tests using the one data file with modifications such as like that performed in this function.
```
Comment (lbodor) [inline src/client/app/shared/jsonix/jsonix.service.spec.ts]: Inappropriate comment. It does not describe what the function does.

File: src/client/app/shared/json-data-view-model/data-view-translator.ts [hunk 6]
```diff
@@ -92,10 +114,95 @@ export class DataViewTranslatorService {
...
+     * @return the translated value for the top-level CartesianPosition or GeodeticPosition
+     */
+    static translateGeodeticPosition(source: any, viewToDataTranslateOptions?: { viewToData: boolean }): any {
+        let mapper = createMapper({alwaysTransform: true, alwaysSet: true});
+        if (viewToDataTranslateOptions
+            && viewToDataTranslateOptions.hasOwnProperty('viewToData')
+            && !viewToDataTranslateOptions['viewToData']) {
+            // data to view translate
+            if (!source || (typeof source === 'object' && !source.hasOwnProperty('point'))) {
+                // ! source happens when the CartesianPosition or GeodeticPosition elements dont exist
```
Comment (lbodor) [inline src/client/app/shared/json-data-view-model/data-view-translator.ts]: I suspect this comment is not true. It probably should not mention Cartesian position.

File: src/client/app/shared/json-data-view-model/data-view-translator.ts [hunk 7]
```diff
@@ -92,10 +114,95 @@ export class DataViewTranslatorService {
         _.merge(target, result);
     }
 
+    /**
+     * CartesianPosition and GeodeticPosition are Point type with 3 values, and it can be optional and due to the XML schema, we achieve
+     * this by having do data for the Positions.  Here are the mappings:
```
Comment (lbodor) [inline src/client/app/shared/json-data-view-model/data-view-translator.ts]: typo: 'do'

File: src/client/app/shared/json-data-view-model/data-view-translator.ts [hunk 8]
```diff
@@ -92,10 +114,95 @@ export class DataViewTranslatorService {
...
+            value.push(source.cartesianPosition_y);
+            value.push(source.cartesianPosition_z);
+            mapper.map('value').to('point.pos.value');
+            return mapper.execute({'value': value});
+        }
+    }
+
+    /**
+     * CartesianPosition and GeodeticPosition are Point type with 3 values, and it can be optional and due to the XML schema, we achieve
+     * this by having do data for the Positions.  Here are the mappings:
```
Comment (lbodor) [inline src/client/app/shared/json-data-view-model/data-view-translator.ts]: typo 'do'

File: src/client/app/shared/json-data-view-model/data-view-translator.spec.ts (hunk 1 above)
Comment (oehm-smith) [inline src/client/app/shared/json-data-view-model/data-view-translator.spec.ts]: Its a .spec file.  Though these are no longer needed and I'd forgot to delete them (mmm, I thought linting picked this up - I'll look at that).

File: src/client/app/shared/json-data-view-model/data-view-translator.ts [hunk 9]
```diff
@@ -92,10 +114,95 @@ export class DataViewTranslatorService {
...
+            value.push(source.cartesianPosition_x);
+            value.push(source.cartesianPosition_y);
+            value.push(source.cartesianPosition_z);
+            mapper.map('value').to('point.pos.value');
+            return mapper.execute({'value': value});
+        }
+    }
+
+    /**
+     * CartesianPosition and GeodeticPosition are Point type with 3 values, and it can be optional and due to the XML schema, we achieve
```
Comment (lbodor) [inline src/client/app/shared/json-data-view-model/data-view-translator.ts]: This function applies only to geodetic position.

File: src/client/app/shared/json-data-view-model/data-view-translator.ts [hunk 10]
```diff
@@ -92,10 +114,95 @@ export class DataViewTranslatorService {
         _.merge(target, result);
     }
 
+    /**
+     * CartesianPosition and GeodeticPosition are Point type with 3 values, and it can be optional and due to the XML schema, we achieve
```
Comment (lbodor) [inline src/client/app/shared/json-data-view-model/data-view-translator.ts]: This function applies only to Cartesian position.

File: src/client/app/shared/jsonix/jsonix.service.spec.ts [hunk 11]
```diff
@@ -287,5 +288,30 @@ export function main() {
       expect(geodesyMl).toContain('!!!SOME INTERESTING REMARKS!!!');
     });
 
+      describe('Undefining Location.cartesianPosition but not Location.geodeticPosition', () => {
+          it('should parse valid Json', () => {
+              let json: Object = modify_undefine_Location_geodeticPosition(JsonServiceSpecData.data());
+
+              let geodesyMl: string = jsonixService.jsonToGeodesyML(json);
+              // console.log('geodesyMl:\n', geodesyMl);
```
Comment (lbodor) [inline src/client/app/shared/jsonix/jsonix.service.spec.ts]: Remove commented code.

File: src/client/app/site-log/site-location.component.ts [hunk 12]
```diff
@@ -44,6 +44,8 @@ export class SiteLocationComponent implements OnInit {
     private miscUtils: any = MiscUtils;
 
     private siteLocationForm: FormGroup;
+    private cartesianPosition: FormGroup;
+    private geodeticPosition: FormGroup;
```
Comment (lbodor) [inline src/client/app/site-log/site-location.component.ts]: The convention to date is to suffix the above two field names with `Form`.

File: src/client/app/site-log/site-location.component.ts [hunk 13]
```diff
@@ -65,16 +67,28 @@ export class SiteLocationComponent implements OnInit {
     }
 
     private setupForm() {
+        this.cartesianPosition = this.formBuilder.group({
+            cartesianPosition_x: ['', Validators.required],
```
Comment (lbodor) [inline src/client/app/site-log/site-location.component.ts]: No underscores in identifier names please.

File: src/client/app/site-log/site-location.component.ts [hunk 14]
```diff
@@ -85,5 +99,51 @@ export class SiteLocationComponent implements OnInit {
...
+        });
+
+        // If any group Item form field has a value then all are required
+        Object.keys(groupItems).forEach((key) => {
+            let itemControl: AbstractControl = positionGroup.controls[key];
+            if (!itemControl) {
+                throw new Error(`Control for group item: ${key} doesnt exist in positionGroup: ${positionGroup}`);
+            }
+            if (someFieldHasValue) {
+                if (! itemControl.validator) {
```
Comment (lbodor) [inline src/client/app/site-log/site-location.component.ts]: No space after !

File: src/client/app/shared/jsonix/jsonix.service.spec.ts [hunk 15]
```diff
@@ -287,5 +288,30 @@ export function main() {
       expect(geodesyMl).toContain('!!!SOME INTERESTING REMARKS!!!');
     });
 
+      describe('Undefining Location.cartesianPosition but not Location.geodeticPosition', () => {
+          it('should parse valid Json', () => {
+              let json: Object = modify_undefine_Location_geodeticPosition(JsonServiceSpecData.data());
+
+              let geodesyMl: string = jsonixService.jsonToGeodesyML(json);
```
Comment (lbodor) [inline src/client/app/shared/jsonix/jsonix.service.spec.ts]: Capital `l` in `geodesyML` please.

File: src/client/app/shared/json-data-view-model/data-view-translator.ts (hunk 2 above)
Comment (oehm-smith) [inline src/client/app/shared/json-data-view-model/data-view-translator.ts]: No won't work.  One of my latest change was moving the code above that.  If the `siteLocation.approximatePositionITRF` is "empty" (it doesn't contain `positions` but only a `TYPE_NAME`) then we need to catch this and create empty positions.  It is made empty when the positions have null values upon save.

File: src/client/app/shared/json-data-view-model/data-view-translator.spec.ts (hunk 1 above)
Comment (oehm-smith) [inline src/client/app/shared/json-data-view-model/data-view-translator.spec.ts]: Yep we are missing the linting rule.  I will do that in a later PR.

File: src/client/app/shared/json-data-view-model/data-view-translator.ts (hunk 3 above)
Comment (oehm-smith) [inline src/client/app/shared/json-data-view-model/data-view-translator.ts]: I'll fix this now though the approach was to do a format commit after you are happy with this.

File: src/client/app/shared/json-data-view-model/data-view-translator.ts (hunk 6 above)
Comment (oehm-smith) [inline src/client/app/shared/json-data-view-model/data-view-translator.ts]: It is.  As we discussed the `siteLocation.approximatePositionITRF` only contains a `TYPE_NAME` and the CP and GP elements dont exist.  I'll update it to include this new information.

Comment (lbodor) [inline src/client/app/shared/json-data-view-model/data-view-translator.ts]: This function has nothing to do with Cartesian position.

File: src/client/app/site-log/site-location.component.ts (hunk 13 above)
Comment (oehm-smith) [inline src/client/app/site-log/site-location.component.ts]: Done in separate PR and rebased. :smile:

File: src/client/app/site-log/site-location.component.ts [hunk 16]
```diff
@@ -65,16 +66,28 @@ export class SiteLocationComponent implements OnInit {
     }
 
     private setupForm() {
+        this.cartesianPositionForm = this.formBuilder.group({
+            cartesianPosition_x: ['', Validators.required],
+            cartesianPosition_y: ['', Validators.required],
+            cartesianPosition_z: ['', Validators.required]
```
Comment (lbodor) [inline src/client/app/site-log/site-location.component.ts]: No underscores please.

File: src/client/app/site-log/site-location.component.ts [hunk 17]
```diff
@@ -65,16 +66,28 @@ export class SiteLocationComponent implements OnInit {
...
+        });
+        this.cartesianPositionForm.valueChanges.subscribe(
+            (change: any) => this.handleLocationPositionGroupChange(change, this.cartesianPositionForm)
+        );
+        this.geodeticPositionForm = this.formBuilder.group({
+            geodeticPosition_lat: [''],
+            geodeticPosition_long: [''],
+            geodeticPosition_height: ['']
```
Comment (lbodor) [inline src/client/app/site-log/site-location.component.ts]: Remove underscores.
